### ISSUE TITLE
Too broad packages picked up when using nohoist option

### ISSUE THREAD

**Issue by maoueh** [yarnpkg/yarn#5579]
**Do you want to request a *feature* or report a *bug*?**
*bug*

**What is the current behavior?**
I started to play with the `nohoist` option on one of our internal project which has a structure like:

```
├── modules
│   ├── client
│   ├── domain
│   └── frontend
│       ├── common
│       └── mobile
```

When adding the `nohoist` option, either in the child projet `modules/frontend/mobile` or at the top level, it appears that on a second `yarn install` call, Yarn is picking up too much things probably looking inside `node_modules` where it should not leading to this series of messages:

```
yarn install v1.5.1
warning Missing name in workspace at "/private/tmp/yarn-1.5.1-too-broad-packages-with-nohoist/packages/frontend/mobile/node_modules/metro-source-map/src", ignoring.
warning Missing name in workspace at "/private/tmp/yarn-1.5.1-too-broad-packages-with-nohoist/packages/frontend/mobile/node_modules/metro/src/ModuleGraph/node-haste", ignoring.
warning Missing name in workspace at "/private/tmp/yarn-1.5.1-too-broad-packages-with-nohoist/packages/frontend/mobile/node_modules/metro/src/Server/symbolicate", ignoring.
error An unexpected error occurred: "There are more than one workspace with name \"nw-pre-gyp-module-test\"".
info If you think this is a bug, please open a bug report with the information provided in "/private/tmp/yarn-1.5.1-too-broad-packages-with-nohoist/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

**If the current behavior is a bug, please provide the steps to reproduce.**

```
git clone https://github.com/maoueh/yarn-1.5.1-too-broad-packages-with-nohoist.git
cd yarn-1.5.1-too-broad-packages-with-nohoist.git
yarn install # This one succeed
yarn install # This one fails with the above messages
```

**What is the expected behavior?**
No warning and no error on the second yarn install

**Possible workaround?**
Narrowing the `workspaces/packages` option resolves the issue. Here's what I used in my project:

```
  "workspaces": {
    "packages": [
      "modules/client",
      "modules/domain",
      "modules/frontend/*",
      "modules/tools/*"
    ],
    "nohoist": ["**/react-native*", "**/react-native*/**"]
  },
```

**Please mention your node.js, yarn and operating system version.**

 * Yarn 1.5.1
 * Node 8.4.0
 * Mac OS X High Sierra

**Comment by connectdotz** [comment]
@maoueh your original workspaces.packages config in the repo is not correct: `"packages": ["modules/**/*"]`. This will indicate domain, **frontend**, common and mobile are all workspaces, which frontend is clearly not. Like you have discovered, you need to provide correct package location, such as `"packages": ["modules/domain", "module/frontend/*"]`

yarn is working as expected once you correctly tell it where your workspaces are.

**Comment by maoueh** [comment]
@connectdotz That's seems wrong as without the `nohoist`, it works correctly and does not raise the issue.

I have a current project using Yarn 1.5.1 that has:

```
  "workspaces": [
    "packages/**/*"
  ],
```

And it works as expected.

My understanding of workspaces and `packages` glob `modules/**/*` is that for a tree looking like:

```
├── modules
│   ├── client
│   │   └── package.json
│   ├── domain
│   │   └── package.json
│   └── frontend
│       ├── common
│       │   └── package.json
│       └── mobile
│           └── package.json
```

Yarn will pick `modules/client/package.json`, `modules/domain/package.json`, `modules/frontend/common/package.json` and `modules/frontend/mobile/package.json` as being package of the workspace without going inside the `node_modules` folder. I already reported similar issue when the initial workspace feature was introduced (https://github.com/yarnpkg/yarn/issues/3986).

The problem with the `modules/**/*` glob arose only when I started adding `nohoist` directives to ensure `react-native` remains under the `modules/frontend/mobile`'s `node_modules`.

In one of the project I have, that uses `modules/**/*`, here the result of doing `ls node_modules | grep acme` (where `acme` is the project suffix I added to all `package.json#name` field):

```
$ yarn install
$ ls node_modules | grep acme
acme-domain
acme-backend-account
acme-backend-common
acme-backend-stores
acme-client
acme-frontend-common
acme-frontend-mobile
acme-frontend-web
```

Am I making wrong assumptions?

**Comment by connectdotz** [comment]
hmmm... I see where you are coming from. I should have given a more completed answer.

Looking at the code, it appeared the fix addressing #3986 was not fully correct. It only ignored **shallow** pattern such as `node_module/*/package.json`. 

On the first run, all is well, because there is nothing under workspace packages, the pattern `modules/**/*` will only find 3 packages as expected.

On the 2nd run, because nohoist has populated the node_modules under these workspace packages, the glob pattern `modules/**/*` will match everything **deeper than one layer**, such as `modules/frontend/mobile/node_modules/node-pre-gyp/lib/util/nw-pre-gyp`. thus encountered the duplicate error...

why workspaces without nohoist work? Because without nohoist, all modules are hoisted to the root, your workspace packages will not have their own local node_modules populated (actually, in theory, these can still happen if you have conflict versions, but just not in your example), therefore, yarn will not find extra match and thus appear to work...

While the bug is not nohoist specific, it did increase the chance to surface it. We should address this issue, however, it is always recommended that you give precise patterns so your install will be faster and more efficient.

**Comment by maoueh** [comment]
@connectdotz Thanks for the time spent here, I appreciate. I'm using `modules/**/*` just so it's easier for developer to add new packages, then don't need to fiddle with the `workspaces` option.

However, even if I get your explanation, there still some stuff that don't fit the reality I'm facing. In the other project, there is conflicts and there is already many other deep `node_modules` at various location within the tree without triggering the error:

```
$ find . -type d -name "node_modules"
(redacted out `node_modules` inside other `node_modules` for easier reading)
./node_modules
./packages/frontend/web/node_modules (2)
./packages/frontend/common/node_modules (3)
./packages/frontend/mobile/node_modules (58)
./packages/backend/common/node_modules (0)
./packages/backend/account/node_modules (0)
./packages/shared/cli/node_modules (0)
./packages/shared/migration/node_modules (0)
```

In parentheses I added the number of elements in the folder. Note also that `mobile` has some nested `node_modules` further down the initial one.

We agree that the problem should arise in this project which is not the case. It arise only when I add the `nohoist` option in a similar project.

I will try to update the repro Git project I linked to this issue adding simulated conflict along the way so that nested `node_modules` exist even without `nohoist` and I'll report back.

I'm really confident it's related to `nohoist` somehow, I just don't know how exactly yet :)

**Comment by maoueh** [comment]
@connectdotz I've updated the repo to add conflicts so that nested `node_modules` exist. The nested node_modules gives currently:

```
./node_modules
(redacted)
./modules/frontend/common/node_modules
./modules/frontend/common/node_modules/react/node_modules
./modules/frontend/mobile/node_modules
./modules/domain/node_modules
./modules/domain/node_modules/fbjs/node_modules
./modules/domain/node_modules/react/node_modules
```

Without the `nohoist` option, it's possible to run `yarn install` without any problem. When using `"nohoist": ["**/react-native*", "**/react-native*/**"]`, I get this error:

```
yarn install v1.5.1
warning Missing name in workspace at "/private/tmp/yarn-1.5.1-too-broad-packages-with-nohoist/modules/frontend/mobile/node_modules/metro-source-map/src", ignoring.
warning Missing name in workspace at "/private/tmp/yarn-1.5.1-too-broad-packages-with-nohoist/modules/frontend/mobile/node_modules/metro/src/ModuleGraph/node-haste", ignoring.
warning Missing name in workspace at "/private/tmp/yarn-1.5.1-too-broad-packages-with-nohoist/modules/frontend/mobile/node_modules/metro/src/Server/symbolicate", ignoring.
error An unexpected error occurred: "There are more than one workspace with name \"nw-pre-gyp-module-test\"".
info If you think this is a bug, please open a bug report with the information provided in "/private/tmp/yarn-1.5.1-too-broad-packages-with-nohoist/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

If I use instead `"nohoist": ["**/debug*", "**/debug*/**"]`, I get no error, it works as expected.

If I use instead `"nohoist": ["**/react", "**/react/**"]`, no I get a complete different error. Off topic, but nohoist pattern does not seems respected here as `react-native` is found under `modules/frontend/mobile` where if I get it right, it should still be hoisted.

But you know what is 100% strange, is that Yarn complains about a dependency that is not directly listed as one of my dependency and furthermore, I already face this error in my other project that is not using `nohoist` yet. The offending dependency is in `node_modules/react-native` at path `local-cli/core/__fixtures__/files/package.json`.

This would kind of prove your point. I'm really baffled about the exact problem in fact now :)

Error for `"nohoist": ["**/react", "**/react/**"]` pattern:

```
yarn install v1.5.1
[1/4] 🔍  Resolving packages...
warning workspace-aggregator-59541422-9305-41be-94ea-c05e81be3a72 > react-animated > gulp-util@3.0.8: gulp-util is deprecated - replace it, following the guidelines at https://medium.com/gulpjs/gulp-util-ca3b1f9f9ac5
warning workspace-aggregator-59541422-9305-41be-94ea-c05e81be3a72 > react-animated > gulp-babel > gulp-util@3.0.8: gulp-util is deprecated - replace it, following the guidelines at https://medium.com/gulpjs/gulp-util-ca3b1f9f9ac5
warning workspace-aggregator-59541422-9305-41be-94ea-c05e81be3a72 > react-animated > gulp > gulp-util@3.0.8: gulp-util is deprecated - replace it, following the guidelines at https://medium.com/gulpjs/gulp-util-ca3b1f9f9ac5
warning workspace-aggregator-59541422-9305-41be-94ea-c05e81be3a72 > react-animated > gulp-flatten > gulp-util@3.0.8: gulp-util is deprecated - replace it, following the guidelines at https://medium.com/gulpjs/gulp-util-ca3b1f9f9ac5
warning workspace-aggregator-59541422-9305-41be-94ea-c05e81be3a72 > react-animated > run-sequence > gulp-util@3.0.8: gulp-util is deprecated - replace it, following the guidelines at https://medium.com/gulpjs/gulp-util-ca3b1f9f9ac5
warning workspace-aggregator-59541422-9305-41be-94ea-c05e81be3a72 > react-animated > webpack-stream > gulp-util@3.0.8: gulp-util is deprecated - replace it, following the guidelines at https://medium.com/gulpjs/gulp-util-ca3b1f9f9ac5
warning workspace-aggregator-59541422-9305-41be-94ea-c05e81be3a72 > react-animated > babel-core > minimatch@2.0.10: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
warning workspace-aggregator-59541422-9305-41be-94ea-c05e81be3a72 > react-animated > gulp > vinyl-fs > glob-stream > minimatch@2.0.10: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
warning workspace-aggregator-59541422-9305-41be-94ea-c05e81be3a72 > react-animated > gulp > vinyl-fs > glob-stream > glob > minimatch@2.0.10: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
warning workspace-aggregator-59541422-9305-41be-94ea-c05e81be3a72 > react-animated > gulp > vinyl-fs > graceful-fs@3.0.11: please upgrade to graceful-fs 4 for compatibility with current and future versions of Node.js
warning workspace-aggregator-59541422-9305-41be-94ea-c05e81be3a72 > react-animated > gulp > vinyl-fs > graceful-fs > natives@1.1.2: This module relies on Node.js's internals and will break at some point. Do not use it, and update to graceful-fs@4.x.
warning workspace-aggregator-59541422-9305-41be-94ea-c05e81be3a72 > react-animated > gulp > vinyl-fs > glob-watcher > gaze > globule > minimatch@0.2.14: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
warning workspace-aggregator-59541422-9305-41be-94ea-c05e81be3a72 > react-animated > gulp > vinyl-fs > glob-watcher > gaze > globule > glob > minimatch@0.2.14: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
warning workspace-aggregator-59541422-9305-41be-94ea-c05e81be3a72 > react-animated > gulp > vinyl-fs > glob-watcher > gaze > globule > glob > graceful-fs@1.2.3: please upgrade to graceful-fs 4 for compatibility with current and future versions of Node.js
error An unexpected error occurred: "Couldn't find package \"rnpm-plugin-test@*\" required by \"react-native-vector-icons@1.0.0\" on the \"npm\" registry.".
info If you think this is a bug, please open a bug report with the information provided in "/private/tmp/yarn-1.5.1-too-broad-packages-with-nohoist/yarn-error.log".
```

**Comment by connectdotz** [comment]
> I've updated the repo to add conflicts so that nested node_modules exist. 

nested modules do not necessarily trigger this bug, it has to be "**deep**", let me give you an example: The glob pattern `node_modules/*/package.json` will match any regular nested pattern like `A/node_modules/B/node_modules/C/node_modules/project.json`, but it won't match `A/node_modules/B/C/project.json`. The offending modules will have to have this type of "deep" file structure, in respect to the matching pattern, to trigger the bug I mentioned earlier. In your repo, that package is `nw-pre-gyp` in `modules/frontend/mobile/node_modules/node-pre-gyp/lib/util/nw-pre-gyp`, an react-native dependency.

> If I use instead "nohoist": ["**/debug*", "**/debug*/**"], I get no error, it works as expected.

that is because this nohoist pattern will not put `react-native` thus `nw-pre-gyp` in your local node_modules, therefore, will not trigger the bug.

> I already face this error in my other project that is not using nohoist yet. The offending dependency is in node_modules/react-native at path local-cli/core/__fixtures__/files/package.json.

I hope it should be pretty clear now why this failed, with or without nohoist: it doesn't satisfy the shallow pattern: `node_modules/*/package.json`, i.e. you will see the bug.

> This would kind of prove your point. I'm really baffled about the exact problem in fact now :)

I hope you are no longer baffled ;-) If you are still not convinced, you could recreate failure examples without nohoist in the following ways:
1.  the easy way: have a clean repo => run first with nohoist => then run again without nohoist => the 2nd run will fail even without nohoist. 
1. the hard way: without nohoist, have 2 workspaces depends on a different react-native version, so 1 of them will be forced to create a local copy in its node_modules, and you should see a similar error as a result (or 3 workspaces with 3 different versions to see the duplicate workspace error)

> Error for "nohoist": ["**/react", "**/react/**"] pattern:
> ...
> error An unexpected error occurred: "Couldn't find package \"rnpm-plugin-test@*\" required by \"react-native-vector-icons@1.0.0\" on the \"npm\" registry.".

this one is a bit more interesting, there are 2 points:
1. the error is from `modules/frontend/mobile/node_modules/react-native/local-cli/core/__fixtures__/files/package.json` which we already discussed above. 
1. however, you would think that react-native should be hoisted to the root since it is not in the nohoist list, therefore should not have stayed in the local node_modules to trigger the bug... turns out, `react-native` has `react` as a peer dependency, and since `react` is nohoist, so goes react-native. One can argue about this behavior (predated nohoist), but it is certainly beyond the scope of this issue, and frankly not relevant.

To conclude, I have already tested the bug fix in your repo and it worked fine, so you can either wait for it or just fix your package pattern.

**Comment by maoueh** [comment]
@connectdotz After reading everything a second time, now I completely understand your explanation that the initial fix used the glob pattern `node_modules/*/package.json` which works in most cases since usually, most packages when installed looks like `node_modules/<package_name>/package.json` but not always.

I now completely get why I was getting the unknown dependency `rnpm-plugin-test` in my previous project that is not using `nohoist` yet.

> One can argue about this behavior (predated nohoist), but it is certainly beyond the scope of this issue, and frankly not relevant.

Totally agree, not relevant here. However, I'm very happy to understand the reasoning so my comprehension of `nohoist` will be much better with this. I'm ok whatever the implementation does.

In conclusion, thank you very much for you time and the detailed explanation, that was really great to understand the underlying system. Thank you!
